### ISSUE TITLE
MA0016: False positive on 'Prefer returning collection abstraction instead of implementation'

### ISSUE THREAD

**Issue by royberris** [meziantou/Meziantou.Analyzer#379]
**Context**
Rule identifier: _MA0016_
Target Framework: _.NET 6.0_
Language version: _10.0_

**Describe the bug**

False positive on 'Prefer returning collection abstraction instead of implementation'. It is showing this when we can not return an abstraction.

**To Reproduce**

When returning a `Task` with a `List` that has no `async` in an extension method. Our scenario is an extension using the `IQueryable ` and `ToListAsync`

````c#
public static Task<List<Example>> ToCustomListAsync(this IQueryable queryable)
        => queryable.Cast<Example>().AsNoTracking().ToListAsync();
````

There is no way to return `IList` without awaiting the task, which is something I do not want.

**Comment by royberris** [comment]
My suggestion would be to not report this issue when a collection is wrapped in a Task that is not async? No idea of that is something you can work with :D

**Comment by meziantou** [comment]
I don't think this is a false-positive. `ToCustomListAsync` is accessible from outside the assembly and returns a concrete collection type. In this case the method is `async`, but the rule is still the same. If this is undesirable, you can disable the diagnostic using an attribute or directive.

Note that you can also create your own extension method to return an abstract collection (based on [EntityFrameworkQueryableExtensions.cs](https://github.com/dotnet/efcore/blob/767bc93c6e6fca19868f486c7df8f4235bcf53cd/src/EFCore/Extensions/EntityFrameworkQueryableExtensions.cs#L2217))

````
    public static async Task<IReadOnlyList<TSource>> ToReadOnlyListAsync<TSource>(
        this IQueryable<TSource> source,
        CancellationToken cancellationToken = default)
    {
        var list = new List<TSource>();
        await foreach (var element in source.AsAsyncEnumerable().WithCancellation(cancellationToken))
        {
            list.Add(element);
        }

        return list;
    }
````

**Comment by royberris** [comment]
Thanks for clearing it up
